### PR TITLE
Add filters to homepage and navbar autocomplete

### LINKED ISSUE #5057: Add filtering results to home page and navbar autocomplete + redesign
👇 Claude's plan which I partially executed before switching to manual. Kept for context.

# Add Filtering Results to Main Site Autocomplete

## Overview

Enhance the main site autocomplete (`Autocomplete.tsx`) to show filter-based suggestions similar to the search page, including:

- "All data on [topic]" suggestions with filter pills
- Country suggestions with flag icons
- Smart parsing of queries to extract countries/topics as separate suggestions

This will provide a more intuitive search experience by showing structured filter options alongside traditional search results.

## Design Reference

Based on Figma design showing search autocomplete with:

- Current query as suggestion ("poverty china")
- Filter-based suggestions ("All data on Poverty", "China" with flag)
- Mixed results including topic pages, articles, data insights, and charts

### Phase 1: Dependencies and Imports ✅ COMPLETE

- [x] Import filtering utilities from `searchUtils.tsx`:
    - [x] `getAutocompleteSuggestionsWithUnmatchedQuery`
    - [x] `createCountryFilter`, `createTopicFilter`, `createQueryFilter`
    - [x] `getFilterIcon`
- [x] Import filtering types from `searchTypes.ts`:
    - [x] `Filter`, `FilterType`
- [x] Import UI components:
    - [x] `SearchFilterPill` from `SearchFilterPill.tsx`
- [x] Import country data utility:
    - [x] `countriesByName` from `@ourworldindata/utils`

### Phase 2: Data Sources Setup ✅ COMPLETE

- [x] ~~Create static topics list within component by copying `ALL_TOPICS` from `Search.tsx`~~
- [x] ~~Create static country list within component by copying AREA_NAMES from `Search.tsx`~~
- [x] Create `useTopicTagGraph` hook in `hooks.ts` to abstract tagGraph fetching logic
- [x] Use dynamic topics list with `useMemo` hook that traverses tagGraph to find all topics (same logic as `Search.tsx`)
- [x] Use dynamic country list with `useMemo` hook from `tagGraph.children` (same logic as `Search.tsx`)
- [x] Update `SiteNavigation.tsx` to use the new `useTopicTagGraph` hook
- [x] Make a note to refactor this using the hooks provided by the search template PR

### Phase 3: Create FilterBasedSource

- [ ] Create new `FilterBasedSource: AutocompleteSource<BaseItem>`
- [ ] Implement `getItems` function that:
    - [ ] Uses `getAutocompleteSuggestionsWithUnmatchedQuery` to parse query
    - [ ] Returns suggestions for countries, topics, and remaining query
    - [ ] Filters out empty/invalid suggestions
- [ ] Implement `onSelect` function that:
    - [ ] Handles country selections → navigate to `/data` with country filter
    - [ ] Handles topic selections → navigate to `/data` with topic filter
    - [ ] Handles query selections → navigate to `/data` with query parameter
- [ ] Implement `getItemUrl` function for proper URL generation

### Phase 4: UI Templates for Filter Items

- [ ] Create template for country suggestions:
    - [ ] Show country flag icon using `getFilterIcon`
    - [ ] Display country name
    - [ ] Use `SearchFilterPill` component
- [ ] Create template for topic suggestions:
    - [ ] Show "All data on" prefix with chart icon
    - [ ] Display topic as filter pill using `SearchFilterPill`
    - [ ] Handle long topic names with proper truncation
- [ ] Create template for current query as suggestion:
    - [ ] Show search icon
    - [ ] Display query text
    - [ ] Match existing search result styling

### Phase 5: Integration with Existing Autocomplete

- [ ] Update `getSources` function to include `FilterBasedSource`:
    - [ ] Add `FilterBasedSource` as highest priority (show at top)
    - [ ] Maintain existing source order: Algolia → Featured → AllResults
    - [ ] Only show `FilterBasedSource` when query exists
- [ ] Ensure proper source ordering and priority
- [ ] Test interaction between filter suggestions and Algolia results

### Phase 6: Styling Integration

- [ ] Ensure styles work within Algolia autocomplete framework
- [ ] Add custom styles for filter-based suggestions if needed, based on `SearchAutocompleteItemContents.scss`:
    - [ ] Proper spacing and alignment
    - [ ] Icon sizing and positioning
    - [ ] Responsive behavior
- [ ] Test styling across different screen sizes

### Phase 7: Analytics Integration

- [ ] Add analytics tracking for filter-based selections:
    - [ ] Track country filter selections
    - [ ] Track topic filter selections
    - [ ] Track query selections from filter suggestions
- [ ] Ensure existing Algolia analytics continue working
- [ ] Add proper position tracking for mixed results

### Phase 8: Performance Optimization

- [ ] Optimize filtering logic for real-time typing:
    - [ ] Debounce heavy operations if needed
    - [ ] Limit number of suggestions returned
    - [ ] Cache country/topic data
- [ ] Test performance with long queries
- [ ] Test performance with large topic/country datasets

### Phase 9: Error Handling and Edge Cases

- [ ] Handle empty query states gracefully
- [ ] Handle queries with no matches
- [ ] Handle malformed or special character queries
- [ ] Ensure fallback behavior when filtering fails
- [ ] Handle navigation errors gracefully

### Phase 10: Testing and Validation

- [ ] Test basic functionality:
    - [ ] Country detection and navigation ("china", "united states")
    - [ ] Topic detection and navigation ("poverty", "climate change")
    - [ ] Mixed queries ("poverty china", "co2 france")
- [ ] Test edge cases:
    - [ ] Very short queries
    - [ ] Very long queries
    - [ ] Special characters and punctuation
    - [ ] Non-English characters
- [ ] Test integration:
    - [ ] Ensure existing Algolia results still work
    - [ ] Ensure recent searches plugin still works
    - [ ] Ensure keyboard navigation works across all suggestion types
- [ ] Test accessibility:
    - [ ] Screen reader compatibility
    - [ ] Keyboard navigation
    - [ ] ARIA labels and roles

### Phase 11: Documentation and Cleanup

- [ ] Update component documentation
- [ ] Add inline comments explaining complex logic
- [ ] Remove any debug code or console.logs
- [ ] Ensure TypeScript types are properly defined

### Architecture Considerations

- Current `Autocomplete.tsx` uses `@algolia/autocomplete-js` while `SearchAutocomplete.tsx` uses custom hooks
- Need to adapt filtering logic to work with Algolia's source format
- Maintain backward compatibility with existing functionality

### Data Requirements

- ~~Need access to topics list (currently only available in search page context)~~ ✅ Solved with `useTopicTagGraph` hook
- Country data available via `countriesByName()` utility
- ~~Consider caching strategies for performance~~ ✅ Implemented with `useMemo` hooks

### Navigation Patterns

- Country filters: `/data?countries=China`
- Topic filters: `/data?topics=Poverty`
- Query searches: `/data?q=poverty+china`
- Maintain consistency with existing search page URL patterns

## Success Criteria

- [ ] Users can see country suggestions with flags when typing country names
- [ ] Users can see "All data on [topic]" suggestions when typing topic names
- [ ] Mixed queries like "poverty china" show both country and topic filter options
- [ ] All suggestions navigate to appropriate search results
- [ ] Existing Algolia search results continue to work
- [ ] Performance remains acceptable with real-time filtering
- [ ] Styling is consistent with design system

## Related Files

- `site/search/Autocomplete.tsx` (main file to modify)
- `site/search/SearchAutocomplete.tsx` (reference implementation)
- `site/search/SearchFilterPill.tsx` (UI component to reuse)
- `site/search/searchUtils.tsx` (utility functions to import)
- `site/search/searchTypes.ts` (type definitions)
- `site/hooks.ts` (new `useTopicTagGraph` hook added)

### PR DESCRIPTION
**Pull request by mlbrgl**
resolves #5057 

## Context

This PR enhances the search autocomplete experience on the homepage and navbar with improved styling and functionality. It introduces a new filters source that suggests topics and countries based on user input, and reshapes the autocomplete UI to be more compact.

🎨 [Figma](https://www.figma.com/design/kxTQYvNWSe8Fx37iOLNG8i/Auto-complete?node-id=40-3803&m=dev)

## Screenshots / Videos / Diagrams

| Homepage | Navbar |
|--------|--------|
|  | | 

## Testing guidance

1. Open the site and click on the search box in the navigation
2. Test typing various search terms to see the new autocomplete suggestions
3. Verify that topic and country suggestions appear appropriately
4. Check that the styling is consistent across all autocomplete items
5. Test on mobile and desktop viewports to ensure responsive behavior
6. Verify that clicking on suggestions navigates to the correct search results page
7. Repeat point above with the homepage search box

- [x] Does the staging experience have sign-off from product stakeholders? -> yes (call Joe) but we might need to address some of the sub-issues from #5092 before merging (see [threads](https://owid.slack.com/archives/C08M41DRVHR/p1751537800437429))

### Before merging
- [x] Google Analytics events were adapted to fit the changes in this PR
- [x] Changes to CSS/HTML were checked on Desktop and Mobile Safari at all three breakpoints (see https://github.com/owid/owid-grapher/issues/5073)
- [x] Changes to HTML were checked for accessibility concerns: yes although we can only go so far without taking Algolia's Autocomplete apart going through a heavy rewrite.

### CODE REVIEW

**Comment by mlbrgl** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/5071?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#5244** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/5244?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5071** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/5071?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/5071?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#4999** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4999?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 39 other dependent PRs ([#5074](https://github.com/owid/owid-grapher/pull/5074) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/5074?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#5083](https://github.com/owid/owid-grapher/pull/5083) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/5083?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#5091](https://github.com/owid/owid-grapher/pull/5091) <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/5091?utm_source=stack-comment-icon" target="_blank">Graphite</a> and 36 others)
* **#4867** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/4867?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#5005** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/5005?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by mlbrgl** [comment]
Still a long way before this gets merged, and PRs will be stacked before it from Martin and Sophia, so I'll update bundlemon closer to the merge if it remains the tip of the stack

**Comment by mlbrgl** [comment]
@ikesau this one is good for review. The [autocomplete feature will receive more attention in the upcoming cycle](https://github.com/owid/owid-grapher/issues/5092) but this is a good stepping stone.

File: site/search/searchHooks.ts [hunk 1]
```diff
@@ -185,3 +187,21 @@ export function useInfiniteSearch<T extends SearchResponse<U>, U>({
         totalResults,
     }
 }
+export const useTopicTagGraph = () => {
```
Comment (ikesau) [inline site/search/searchHooks.ts:190]: On the home page, this is called twice and the JSON is fetched twice (once from Autocomplete and once from SiteNavigation)

It'll be cached on prod so this is probably fine. Just flagging it 🙂

File: site/search/Autocomplete.tsx [hunk 2]
```diff
@@ -196,70 +197,141 @@ const AlgoliaSource: AutocompleteSource<BaseItem> = {
...
-                    <div className="aa-ItemContent">
-                        <div className="aa-ItemIcon">
-                            <FontAwesomeIcon icon={faSearch} />
-                        </div>
-                        <div className="aa-ItemContentBody">
-                            {item.title as string}
-                        </div>
-                    </div>
-                </div>
+                <span className="autocomplete-item-contents">
```
Comment (ikesau) [inline site/search/Autocomplete.tsx]: style nitpick, I feel like it'd be nicer to return the `match` call directly, where each match returns `<span className="autocomplete-item-contents">` instead of a fragment.

e.g.

```tsx
return match(filter.type)
  .with(FilterType.QUERY, () => (
    <span className="autocomplete-item-contents">
    // etc...
```

File: site/search/Autocomplete.tsx [hunk 3]
```diff
@@ -196,70 +197,141 @@ const AlgoliaSource: AutocompleteSource<BaseItem> = {
...
-        return [
-            {
-                slug: `/search${queryParamsToStr({ q: query })}`,
-                title: `All search results for "${query}"`,
-            },
-        ]
-    },
+        if (!query.trim()) return []
 
+        const suggestions = getAutocompleteSuggestionsWithUnmatchedQuery(
```
Comment (ikesau) [inline site/search/Autocomplete.tsx:262]: I'm sure you don't want to dive into `searchWithWords` again, but when `limit` is 1, it's a bit surprising that the countries aren't alphabetized (i.e. I'd expect Cambodia to be the result here)

image

Not a blocker for me, but worth checking with Joe I guess?

File: site/search/searchUtils.tsx [hunk 4]
```diff
@@ -705,3 +718,54 @@ export async function fetchJson<TResult>(url: string): Promise<TResult> {
...
+        .with(
+            WordpressPageType.Other,
+            OwidGdocType.Author, // Should never be indexed
+            OwidGdocType.Fragment, // Should never be indexed
+            OwidGdocType.Homepage, // Should never be indexed
+            () => ({ name: "", icon: faFileLines })
+        )
+        .exhaustive()
+}
+export const SEARCH_BASE_PATH = "/data"
```
Comment (ikesau) [inline site/search/searchUtils.tsx:1175]: Just checking - is it planned to create a redirect from /search to /data ?

File: site/search/searchUtils.tsx [hunk 5]
```diff
@@ -705,3 +718,54 @@ export async function fetchJson<TResult>(url: string): Promise<TResult> {
...
 }
+
+export const getUrlParamNameForFilter = (filter: Filter): SearchUrlParam =>
+    match(filter.type)
+        .with(FilterType.COUNTRY, () => SearchUrlParam.COUNTRY)
+        .with(FilterType.TOPIC, () => SearchUrlParam.TOPIC)
+        .with(FilterType.QUERY, () => SearchUrlParam.QUERY)
+        .exhaustive()
+
+export const getItemUrlForFilter = (filter: Filter): string => {
```
Comment (ikesau) [inline site/search/searchUtils.tsx:1133]: An example comment would be helpful here 🙂

**Review by ikesau — APPROVED** [review]

File: site/search/Autocomplete.tsx (hunk 3 above)
Comment (mlbrgl) [reply site/search/Autocomplete.tsx:262]: good point, thanks for raising it. As I understand it, `fuzzysort` will score shorter words higher, and I didn't see a way to override that behaviour natively.

Screenshot 2025-08-18 at 16 05 10

It's also not clear whether alphabetical order is the more valuable criteria here. If we were to go down this route, population might be for instance a better proxy for importance.

Either way I'll bring it up, likely as a limitation for now.

File: site/search/searchUtils.tsx (hunk 4 above)
Comment (mlbrgl) [reply site/search/searchUtils.tsx:1175]: I think we'll probably settle for bringing everything under `/search` but I'll draft a more concrete plan in #5233

File: site/search/searchUtils.tsx (hunk 5 above)
Comment (mlbrgl) [reply site/search/searchUtils.tsx:1133]: good call, thanks 👍

File: site/search/searchHooks.ts (hunk 1 above)
Comment (mlbrgl) [reply site/search/searchHooks.ts:190]: yes, and this is happening on virtually every page, since both site nav and autocomplete are present there too. That makes a compelling argument to try and remove the second round trip through a service worker, a react query set up or  inlining the json. But it is also probably ok at this point given that the content of neither of those component is visible on page load and requires user interaction, so the validation delay (the only thing we can really optimize here given browser caching is indeed already doing the heavy lifting of preventing the content from being downloaded twice) will be in regular network conditions always "absorbed" by the user reaction time (ms vs seconds).

**Comment by mlbrgl** [comment]
folded into #4999
